python/ray/data/impl/push_based_shuffle.py

```diff
@@ -0,0 +1,499 @@
+        return self._merge_task_options[merge_idx]
+
+    def _compute_reducers_per_merge_task(self, output_num_blocks):
+        """
+        Each intermediate merge task will produce outputs for a partition of P
+        final reduce tasks. This helper function computes P based on the merge
+        task index.
+        """
+        self.num_reducers_per_merge_idx = []
+        self.reducer_idx_to_merge_idx = []
+        for merge_idx in range(self.num_merge_tasks_per_round):
+            partition_size = output_num_blocks // self.num_merge_tasks_per_round
+            extra_blocks = output_num_blocks % self.num_merge_tasks_per_round
+            if merge_idx < extra_blocks:
+                partition_size += 1
+
+            self.num_reducers_per_merge_idx.append(partition_size)
+            for _ in range(partition_size):
+                if len(self.reducer_idx_to_merge_idx) < output_num_blocks:
+                    self.reducer_idx_to_merge_idx.append(merge_idx)
+
+
+class PushBasedShufflePlan(ShuffleOp):
+    """
+    Push-based shuffle merges intermediate map outputs on the reducer nodes
+    while other map tasks are executing. The merged outputs are merged again
+    during a final reduce stage. This works as follows:
+
+    1. Submit rounds of concurrent map and merge tasks until all map inputs
+    have been processed. In each round, we execute:
+
+       M map tasks
+         Each produces N outputs. Each output contains P blocks.
+       N merge tasks
+         Takes 1 output from each of M map tasks.
+         Each produces P outputs.
+       Where M and N are chosen to maximize parallelism across CPUs. Note that
+       this assumes that all CPUs in the cluster will be dedicated to the
+       shuffle job.
+
+       Map and merge tasks are pipelined so that we always merge the previous
+       round of map outputs while executing the next round of map tasks.
+
+    2. In the final reduce stage:
+       R reduce tasks
+         Takes 1 output from one of the merge tasks from every round.
+
+    Notes:
+        N * P = R = total number of output blocks
+        M / N = merge factor - the ratio of map : merge tasks is to improve
+          pipelined parallelism. For example, if map takes twice as long to
+          execute as merge, then we should set this to 2.
+        See paper at https://arxiv.org/abs/2203.05072 for more details.
+    """
+
+    def execute(
+        self,
+        input_blocks: BlockList,
+        output_num_blocks: int,
+        clear_input_blocks: bool,
+        *,
+        map_ray_remote_args: Optional[Dict[str, Any]] = None,
+        reduce_ray_remote_args: Optional[Dict[str, Any]] = None,
+        merge_factor: int = 2,
+    ) -> Tuple[BlockList, Dict[str, List[BlockMetadata]]]:
+        logger.info("Using experimental push-based shuffle.")
+        # TODO(swang): For jobs whose reduce work is heavier than the map work,
+        # we should support fractional merge factors.
```

Comment:
I don't understand what fractional merge factors mean. Like, how would a merge task merge 1.5x map outputs? And why is this related with reduce/map workload ratio?

Reply:
It just means you have more merge tasks than map tasks. So if you have map:merge = 1/2 and 10 map tasks for every round, each merge task will output 1/2 of a partition's worth of data. Presumably you'd also want to have many more reduce tasks than map tasks in this scenario.

But yeah, there's no evidence yet that we'd want to support this case. I haven't seen a case yet where the reduce-side computation is heavier than the map.